### ISSUE TITLE
Feature request: frame spacing

### ISSUE THREAD

**Issue by Verubato** [Verubato/frame-sort#6]
Add a feature to configure some padding/spacing between frames.
e.g. bottom padding of 10 would add 10 points of padding at the bottom of each frame to add some spacing.

**Comment by Verubato** [comment]
What approach did you take to achieve this? Did you modify the values passed to `SetPoint()`?

**Comment by XyzKangUI** [comment]
> What approach did you take to achieve this? Did you modify the values passed to `source.Frame:SetPoint(...)`?

Yes, I did. Here is how i've done it:

```
local spacing = {
    player = 10,
    party1 = 0,
    party2 = -10,
    party3 = -20,
    party4 = -30,
}

function addon:ShuffleFrames(orderedUnits, framesByUnit, framesByIndex)
    -- probably too complicated to calculate positions due to the whole flow container layout logic
    -- so instead we can just re-use the existing positions and shuffle them
    -- probably safer and better supported this way anyway
    for i = 1, #orderedUnits do
        local sourceUnit = orderedUnits[i]
        local source = framesByUnit[sourceUnit]
        local target = framesByIndex[i]

        assert(source ~= nil)
        assert(target ~= nil)

        source.Frame:ClearAllPoints()

        for j = 1, #target.Points do
            local point = target.Points[j]

            local unitSpacing = spacing[sourceUnit] or 0

            point[5] = point[5] + unitSpacing
            source.Frame:SetPoint(unpack(point))
        end
    end
end
```

**Comment by Verubato** [comment]
Thanks for the code snippet!

Indeed it's going to be a bit tricky, especially for raids. 

Think I'd also need to handle:
- Multiple sorting calls to not apply spacing more than once.
- Different raid frames sizes, e.g. 3 or 4 frames per row instead of the usual 5.
- Config changes (e.g. role/group/alphabetical) to remove previous spacing and apply the new.
- On the fly changes to the spacing step config values.

Bit harder than I originally thought.

**Comment by XyzKangUI** [comment]
A few notes:
1. It would be 'wise' to hide the CompactPartyFrame sliders for wrath users
2. The assertion on pets throws an error all the time - maybe add a raidOptionDisplayPets cvar check?
3. The spacing doesn't persist through reload.

**Comment by Verubato** [comment]
Thanks for the preliminary testing and feedback.
I've got Retail working pretty nicely but it seems the same code doesn't work well at all for Wotlk.
I'll get Wotlk working and then figure out what code can be made reusable between the two.

**Comment by XyzKangUI** [comment]
Should be fine. Only scenario i can think of a taint is when the SetPoint is called in combat.

**Comment by Verubato** [comment]
Nah it's fine, you've been very helpful to me.

**Comment by Verubato** [comment]
I've made further progress on this with spacing added to Wotlk.

If you have some time it'd be great if you can take a look and let me know what issues you see. I've only been able to test with 3 members on the PTR server, so I'm unsure how it goes with say 10+ members (do you have any way to check?).

Known issues from me are:
* Pets spacing not working properly in Wotlk when "Keep Groups Together" == true.
* Spacing not working in Wotlk when in a party (not raid) and "Keep Groups Together" == true.
* When moving raid members around to different groups, the spacing isn't applied.

**Comment by XyzKangUI** [comment]
When keep groups together is enabled it will convert the frames to CompactPartyFrame / CompactRaidGroup(?). I believe you cannot sort these in Wrath. Anyway ill see what happends with 10+ members soon, will keep you posted in aprox 1-2h.

**Comment by XyzKangUI** [comment]
When you show more than 5 frames (e.g. in a raid with multiple groups) it will not do the 'vertical' sorting while horizontal still works. However just enabling 1 group and hiding the other groups will restore the vertical sorting functionality, while it breaks the horizontal sorting.

**Comment by XyzKangUI** [comment]
I also notice now that pet frames somehow create a gap after the sorting been applied?

**Comment by Verubato** [comment]
Thanks for that.
I think I've got it working fairly nicely now with the exception of pets.

**Comment by Verubato** [comment]
Pets are becoming very painful to get working and there's still some fixes/work I need to do in that area.
I'm thinking of making a release without pet spacing as it might take me a while to figure that out.

I'd be grateful if you could try out the latest changes and see if you discover any non-pet issues?

**Comment by Verubato** [comment]
Published release 3.3.0 with this feature in it.
I'll keep this ticket open for a little bit longer in case XyzKang find's any bugs/issues.
Regarding pets, I'll working on getting them working nicely eventually.

**Comment by XyzKangUI** [comment]
It is looking great, only one thing I've noticed is that when you hide groups (e.g. 2-8) it requires an update. Great job btw.
